### PR TITLE
Bug: "infinite loop" in glb/3

### PR DESCRIPTION
**Pull request by gomoripeti**
I did not manage to track down the root cause so I only provide a
negative example, when `glb/3` call takes a lot of time. (I'm not sure if
it will ever return).

Someone with better shrinking abilities could help solve this mystery ;)

### CODE REVIEW

**Comment by UlfNorell** [comment]
The problem isn't `glb` but `deep_normalize`, which loops on recursive types. It's only used by the tests to let us compare types more easily, so this should never be an issue for actual type checking.

**Comment by gomoripeti** [comment]
Actually I hit this first during self-gradualization (on this branch https://github.com/josefs/Gradualizer/pull/128) and also reproduced it when called `glb/3` directly from the shell. It is true the test-case I committed here has the `deep_normalize` problem too, but couldn't be that `glb` also has an issue with recursively normalizing?

**Comment by gomoripeti** [comment]
I pushed a new commit which has the problematic type that I saw tracing self-gradualization. It has user_types not exported from erl_parse, so I need to put the actual AST in there.

**Comment by Zalastax** [comment]
known_problems are run in the test suite so adding an infinite loop to known_problems would cause the test suite to timeout, which would raise an eunit failure. I suggest we leave to PR open like this instead.

**Comment by gomoripeti** [comment]
My intention was also not to merge this PR, just to provide a runable example. It should rather be viewed as an issue ticket. It should be closed without merge when the original issue is resolved. Hope someone can track this down.

**Comment by gomoripeti** [comment]
Update: I added a glb cache which stores GlbType for already calculated type pairs and...surprise, the negative test case finishes quickly. So it seems this was not an infinite loop, just slowness as `abstract_type()` is a really wide recursive union.

Now the question is, how to implement this properly? (currently using proc dict)
- One option is to have a glb cache in gradualizer_db. (Should it get invalidated? I think currently also gradualizer_db is not updated when a module is reloaded - this could happen in a setup where gradualizer is constantly running and an IDE talks to it to give one-the-fly check results. OK, one thing at a time, no module reload for now)
- Some pure implementation where the `A` map is passed around, it must be properly updated and propagated between sibling branches of unions, tuples etc, not just downward.

@josefs @Zalastax @zuiderkwast @UlfNorell what do you think?

File: src/typechecker.erl [hunk 1]
```diff
@@ -467,7 +494,15 @@ glb_ty(Ty1 = {type, _, tuple, Tys1}, Ty2 = {type, _, tuple, Tys2}, A, TEnv) ->
 						  glb(T1, T2, A, TEnv)
 					  end,
 					  Tys1, Tys2)),
-	    {type(tuple, Tys), constraints:combine(Css)}
+            TupleType =
+                case lists:any(fun(?type(none)) -> true; (_) -> false end,
```
Comment (gomoripeti) [inline src/typechecker.erl]: this should probably be part of normalize?

**Comment by Zalastax** [comment]
I don't think we should need to invalidate the cache for correctness but maybe we need it to keep memory usage at bay.

Passing the cache around manually seems cumbersome. If this was Haskell I would suggest a state monad. I think ETS is what we want here.

**Comment by zuiderkwast** [comment]
I don't think memory usage is an issue. If it is, we can clear the cache after checking each module.

An ETS table or a process seems fine. We could use gradualizer_db, but perhaps it is not good enough separation of concern. It's an interface to other modules for remote types and specs (so perhaps it should be renamed to gradualizer_remote or so).

Maybe a gradualizer_cache module? We may want to cache other things, e.g. subtype lookups, too.

**Comment by Zalastax** [comment]
My hunch is that ETS should be faster than a process so I would favor that. This hunch is based on hearsay though so it is what it is.

**Comment by gomoripeti** [comment]
I agree that the pure solution is cumbersome.
I also thought of an ETS table (to be wrapped in a module API but read/written without process interaction, and only owned by a gen_server process) It is a good idea to put it in a separate module than gradualizer_db.

**Comment by Zalastax** [comment]
Looks good! Great job! Anything else you want to include?

Mental note: consider making an OTP supervisor structure and include gradualizer_cache in it (unless someone protests).

**Comment by gomoripeti** [comment]
I admit, I was just lazy to add the _app and _sup files, but yes with now 2 gen_servers, it's time that Gradualizer can be properly `application:start`-ed

I dont have anything more coming, just to remove the TMP commit from this PR.

File: src/gradualizer_cache.erl [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+%%===================================================================
+
+-spec start_link() -> {ok, Pid :: pid()} |
+                      {error, Error :: {already_started, pid()}} |
+                      {error, Error :: any()}.
+start_link() ->
+    gen_server:start_link({local, ?SERVER}, ?MODULE, [], []).
+
+-spec ensure_started() -> ok | {error, Error :: any()}.
+ensure_started() ->
```
[the remark this answers is not in the document]
Comment (Zalastax) [reply src/gradualizer_cache.erl]: Yes, it should be in a new PR that turns gradualizer into an OTP application!

**Comment by gomoripeti** [comment]
Under pressure I created the OTP app :)

**Pull request merged**